File: backup/backup.go
package backup

import (
	"github.com/timabell/disk-hog-backup/backup_sets"
	"github.com/timabell/disk-hog-backup/dhcopy"
	"github.com/timabell/disk-hog-backup/hard_linker"
	"log"
	"os"
	"path/filepath"
	"time"
)

func Backup(source string, dest string, getTime func () (time.Time)) (setName string, err error) {
	err = os.MkdirAll(dest, os.ModePerm)
	if err != nil {
		log.Fatal(err)
	}
	lastSetName, err := backup_sets.FindLatestSet(dest)
	if err != nil {
		log.Fatalf("Failed to search for previous backup set: %s", err)
	}
	setName, err = backup_sets.CreateEmptySet(dest, getTime)
	if err != nil {
		log.Fatalf("Couldn't create set folder: %s", err)
	}
	destFolder := filepath.Join(dest, setName)
	if lastSetName != "" {
		lastSetPath := filepath.Join(dest, lastSetName)
		hard_linker.HardLinkCopy(lastSetPath, destFolder)
	}
	log.Printf("backing up %v into %v\n", source, destFolder)
	err = dhcopy.CopyFolder(source, destFolder)
	return
}

File: src/backup/backup.rs
use crate::backup_sets::backup_set::create_empty_set;
use crate::dhcopy::copy_folder::copy_folder;
use chrono::Utc;
use std::fs;
use std::io;
use std::path::Path;

pub fn backup(source: &str, dest: &str) -> io::Result<String> {
	fs::create_dir_all(dest)?;
	let set_name = create_empty_set(dest, || Utc::now())?;
	let dest_folder = Path::new(dest).join(&set_name);
	println!("backing up {} into {:?}", source, dest_folder);
	copy_folder(source, dest_folder.to_str().unwrap())?;
	Ok(set_name)
}
